Replaced sequence implementation with finger trees

br4sco opened the pull request:
Here we have changed the underlying implementation of `mexpr` sequences from lists to the finger trees implementation in [https://ocaml-batteries-team.github.io/batteries-included/hdoc2/BatFingerTree.html](https://ocaml-batteries-team.github.io/batteries-included/hdoc2/BatFingerTree.html) to improve the performance of sequence operations.

br4sco committed:
Replaced mexpr seq implementation with finger trees

br4sco committed:
Renamed nth to get and added splitAt

br4sco committed:
Added head, tail, init and last

br4sco committed:
Added set

br4sco committed:
Added snoc

br4sco committed:
Move slice to seq.mc

br4sco committed:
Remove duplicate tests

br4sco committed:
Renamed makeseq to makeSeq

br4sco committed:
Update comments to reflect seqs rather than lists

br4sco committed:
Use last and init in foldr1 to improve performance

br4sco committed:
Update README.md

br4sco committed:
Renaming of helper functions to reflect sequences rather than lists

johnwikman commented on stdlib/seq.mc:
```diff
@@ -1,9 +1,15 @@
-let tail = lam s. slice s 1 (length s)
 let null = lam seq. eqi 0 (length seq)
 
+let slice = lam seq. lam n. lam m.
```
Maybe I am being too pendantic, but could it be an idea to rename the `n` and `m` arguments to `off` (or `offset`) and `cnt` (or `count`) respectively? It would aid readability since `n` and `m` does not tell me very much about what those arguments represent.

br4sco replied:
Sure!

br4sco committed:
Renamed arguments for better readability

br4sco committed:
Replaced sequence patterns implementation with finger trees

br4sco committed:
Make split_at more efficient, move head, tail, init, last to stdlib

split_at is changed to make splitting around fist and last element amortized O(1)

br4sco committed:
define alias for pattern length

EliasC commented on src/boot/ast.ml:
```diff
@@ -172,7 +174,7 @@ and seqMatchType =
 and pat =
 | PatNamed of info * patName                      (* Named, capturing wildcard *)
-| PatSeq   of info * pat list * seqMatchType      (* Sequence pattern *)
+| PatSeq   of info * pat Mseq.t * seqMatchType    (* Sequence pattern *)
```
Should this really be an `Mseq.t`? It seems like this is representing a piece of MExpr syntax and not a runtime value.

br4sco replied:
The main reason for this data-structure is that we apply a `fold_right2` over a sequence of patterns and a sequence when matching over sequences in `mexpr.ml` line 518. If we keep these patterns in a list we would have to write a custom `fold_right2` that folds over a sequence and a list rather than two sequences.

br4sco replied:
Idk what is best

EliasC replied:
That sounds reasonable!

br4sco replied:
I thought about it some more, I think we can change it to a list and then just convert it to a sequence when doing a matching. Since the number of patterns will only be "a handful", speed should not be a factor.

EliasC commented on src/boot/mexpr.ml:
```diff
@@ -141,6 +147,7 @@ let fail_constapp f v fi = raise_error fi ("Incorrect application. function: "
    The reason for this is that if-expressions return expressions
    and not values. *)
 let delta eval env fi c v  =
+    let index_out_of_bounds_in_seq_msg = "Out of bound access in sequence." in
```
"Out of bound**s**"?

br4sco replied:
Ok!

EliasC commented on src/boot/mexpr.ml:
```diff
@@ -481,29 +493,35 @@ let rec debruijn env t =
+    assert (n <= Mseq.length s);
+    if Mseq.length s == 0 then (Mseq.empty, Mseq.empty)
+    else if n < Mseq.length s then Mseq.split_at s n
+    else (s, Mseq.empty)
```
Why are these special cases needed? Is our intended semantics of `split_at` different from the one in the library?

br4sco replied:
Yes, this local function `slitNth` has a different semantics. Less so if we change the semantics of `split_at` so that `split_at [1,2,3] 3` equals `([1,2,3],[])`. Still `splitNth s 0` equals `([],[])` which is not true for `split_at` and I don't think we want `split_at` to have this semantics. In any case, I agree that we should depart from the `split_at [1,2,3] 3` semantics of batteries and change the code a bit to remove `splitNth` altogether.

br4sco replied:
I decided to keep `splitNth` but change its name to reflect the difference in semantics.

EliasC commented on stdlib/seq.mc:
```diff
@@ -1,8 +1,18 @@
 let null = lam seq. eqi 0 (length seq)
+let head = lam seq. get (splitAt seq 1).0 0
+let tail = lam seq. (splitAt seq 1).1
+let last = lam seq. get (splitAt seq (subi (length seq) 1)).1 0
+let init = lam seq. (splitAt seq (subi (length seq) 1)).0
```
Is `splitAt` clever enough to do the efficient thing here?

br4sco replied:
I believe so, see the implementation of `split_at` in `mseq.ml`

EliasC commented on test/mexpr/seq.mc:
```diff
@@ -10,43 +10,45 @@ utest [] with [] in
+// splitAt seq n
+utest splitAt [1,2,3] 0 with ([],[1,2,3]) in
+utest splitAt [1,2,3] 1 with ([1],[2,3]) in
+utest splitAt [1,2,3] 2 with ([1,2],[3]) in
```
I'm surprised that `splitAt [1,2,3] 3` does not equal `([1,2,3, [])` (now it throws an error)! I think this is the behaviour of the finger tree library.

br4sco replied:
`split_at` in batteries has this semantics, but we might want to change this semantics for `splitAt`

br4sco replied:
Changed semantics of `splitAt`

EliasC reviewed (changes requested):
This mostly looks really good! I think the corner case of `splitAt l n` when `n` = `length l` should work though.

EliasC commented:
Another comment is that we should be able to remove the home-rolled `Option` modules and replace them by the one from Batteries, but I can make sure to remember this.

br4sco committed:
Change semantics of splitAt to allow splitting at seq length

br4sco committed:
Simplify and clarify matching on sequences implementation

EliasC commented on src/boot/mexpr.ml:
```diff
@@ -147,7 +147,7 @@ let fail_constapp f v fi = raise_error fi ("Incorrect application. function: "
    and not values. *)
 let delta eval env fi c v  =
-    let index_out_of_bounds_in_seq_msg = "Out of bound access in sequence." in
+    let out_of_bounds_in_seq_msg = "Out of bounds." in
```
Sorry, I was unclear! My comment was that it should be "Out of bounds" and not "out of bound". I think the message should be "Out of bounds access in sequence". Just "out of bounds" is a bit unspecific.

br4sco replied:
Fixed

br4sco committed:
Change out of bounds access in sequences error message

EliasC reviewed (approved):
I'm happy with this now! Well done!

br4sco committed:
Added complexity notes on split_at

br4sco committed:
Re-define ziptWith to make it tail-recursive

br4sco committed:
Simplify slice definition

br4sco committed:
Offset must be within sequence
